README
NAME
    Plack::Middleware::Headers - modify HTTP response headers

SYNOPSIS
      use Plack::Builder;

      my $app = sub {['200', [], ['hello']]};
      builder {
          enable 'Headers',
            set => ['X-Plack-One' => '1'],
            append => ['X-Plack-Two' => '2'],
            unset => ['X-Plack-Three'];
          enable 'Headers',
            code => '404',
            set => ['X-Robots-Tag' => 'noindex, noarchive, follow'];
          $app;
      };

DESCRIPTION
    Plack::Middleware::Headers

AUTHOR
    Masahiro Chiba

CONTRIBUTORS
    Wallace Reis "wreis@cpan.org", Jakob Voß <voss@gbv.>

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    Plack::Middleware Plack::Builder



Changes
Revision history for Perl module Plack::Middleware::Headers

0.10  Sat Sep 14 14:35 UTC 2013
    - fork renamed to Plack::Middleware::Headers
    - migrated to dzil
    - added conditional argument 'when' (Jakob Voß) 
    - added conditional argument 'code' (Wallace Reis)

0.04  Thu Sep  9 06:30 UTC 2010
    - support delayed response (thanks to Olaf Alders)

0.03  Thu Dec 17 07:57 UTC 2009
    - Added parent dependency

0.02  Thu Dec 17 07:51 UTC 2009
    - Added Plack::Middleware dependency

0.01  Thu Dec  3 22:19 2009
    - original version


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "modify HTTP response headers",
  "AUTHOR" => "Jakob Vo\303\237",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Plack-Middleware-Headers",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Plack::Middleware::Headers",
  "PREREQ_PM" => {
    "Plack::Middleware" => 0
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.10",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Plack/Middleware/Header.pm
package Plack::Middleware::Headers;

use strict;
use 5.008_001;
use parent qw(Plack::Middleware);
use Plack::Util::Accessor qw(set append unset code when);

use Plack::Util;
use Scalar::Util qw(reftype);

our $VERSION = '0.06';

sub call {
    my $self = shift; 
    my $res  = $self->app->(@_);

    $self->response_cb(
        $res,
        sub {
            my $res = shift;

            if ($self->code and $self->code ne $res->[0]) {
                return;
            }

            my $headers = $res->[1];

            if ($self->when) {
                my @when  = @{$self->when};
                my $match = 0;
                while (my($key, $check) = splice @when, 0, 2) {
                    my $value = Plack::Util::header_get($headers, $key);
                    if (!defined $check) {            # missing header
                        next if defined $value;     
                    } elsif( ref $check ) {           # regex match header
                        next if $value !~ $check;
                    } elsif ( $value ne $check ) {    # exact header
                        next;
                    }
                    $match = 1; 
                    last;
                }
                return unless $match;
            }
            if ( $self->set ) {
                Plack::Util::header_iter($self->set, sub {Plack::Util::header_set($headers, @_)});
            }
            if ( $self->append ) {
                push @$headers, @{$self->append};
            }
            if ( $self->unset ) {
                Plack::Util::header_remove($headers, $_) for @{$self->unset};
            }
        }
    );
}

1;

__END__

=head1 NAME

Plack::Middleware::Headers - modify HTTP response headers

=head1 SYNOPSIS

  use Plack::Builder;

  builder {
      enable 'Headers',
        set    => ['X-Plack-One' => '1'],
        append => ['X-Plack-Two' => '2'],
        unset  => ['X-Plack-Three'];
      enable 'Headers',
        code   => '404',
        set    => ['X-Robots-Tag' => 'noindex, noarchive, follow'];
      enable 'Headers',
        when   => ['Content-Type' => qr{^text/}],
        set    => ['Content-Type' => 'text/plain'];

      sub {['200', [], ['hello']]};
  };

=head1 DESCRIPTION

Plack::Middleware::Headers

=head1 AUTHOR

Masahiro Chiba

=head1 CONTRIBUTORS

Wallace Reis C<wreis@cpan.org>,
Jakob Voß

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<Plack::Middleware>, L<Plack::Builder>

=encoding utf8

=cut


lib/Plack/Middleware/Headers.pm
package Plack::Middleware::Headers;
#ABSTRACT: modify HTTP response headers

use strict;
use 5.008_001;
use parent qw(Plack::Middleware);
use Plack::Util::Accessor qw(set append unset code when);

use Plack::Util;
use Scalar::Util qw(reftype);

#VERSION

sub prepare_app {
    my $self = shift; 

    if (ref $self->when and ref $self->when eq 'ARRAY') {
        my @when  = @{$self->when};
        $self->when( sub {
            my @headers = @_;
            my $match = 0;
            while (my($key, $check) = splice @when, 0, 2) {
                my $value = Plack::Util::header_get(\@headers, $key);
                if (!defined $check) {            # missing header check
                    next if defined $value;     
                } elsif( !defined $value ) {      # header missing
                    next;
                } elsif( ref $check ) {           # regex match header
                    next if $value !~ $check;
                } elsif ( $value ne $check ) {    # exact header
                    next;
                }
                return 1; 
            }
            return;
        });
    }
}

sub call {
    my $self = shift; 
    my $res  = $self->app->(@_);

    $self->response_cb(
        $res,
        sub {
            my $res = shift;

            if ($self->code and $self->code ne $res->[0]) {
                return;
            }

            my $headers = $res->[1];

            if ($self->when and !$self->when->(@$headers)) {
                return;
            }

            if ( $self->set ) {
                Plack::Util::header_iter(
                    $self->set, sub {Plack::Util::header_set($headers, @_)}
                );
            }
            if ( $self->append ) {
                push @$headers, @{$self->append};
            }
            if ( $self->unset ) {
                Plack::Util::header_remove($headers, $_) for @{$self->unset};
            }
        }
    );
}

1;

__END__

=head1 SYNOPSIS

  use Plack::Builder;

  builder {
      enable 'Headers',
        set    => ['X-Plack-One' => '1'],
        append => ['X-Plack-Two' => '2'],
        unset  => ['X-Plack-Three'];
      enable 'Headers',
        code   => '404',
        set    => ['X-Robots-Tag' => 'noindex, noarchive, follow'];
      enable 'Headers',
        when   => ['Content-Type' => qr{^text/}],
        set    => ['Content-Type' => 'text/plain'];

      sub {['200', [], ['hello']]};
  };

=head1 DESCRIPTION

This L<Plack::Middleware> simplifies creation (C<set> or C<append>), deletion
(C<unset>), and modification (C<set>) of L<PSGI> response headers. The
modification can be enabled based on response code (C<code>) or existing
response headers(C<when>). Use L<Plack::Middleware::Conditional> to enable the
middleware based in I<request> headers.

=head1 CONFIGURATION

=over 4

=item set

Overwrites existent header(s).

=item unset

Remove existing header(s).

=item append

Add header(s).

=item code

Optional HTTP response code that modification of response headers is limited
to.

=item when

Optional check on the response headers that must be true to actually modify
headers. Either one provides a list of headers for which one of them must
match. Matching can be tested against:

    header => undef,    # missing header
    header => $scalar   # exact value
    header => /$regexp/ # regular expression

Alternatively one can check with a code reference that all response headers
are passed to as list.

=back

=head1 CONTRIBUTORS

This module is an extened fork of L<Plack::Middleware::Header>, originally
created by Masahiro Chiba. Additional contributions by Wallace Reis.

=head1 SEE ALSO

L<Plack::Middleware>, L<Plack::Builder>

=encoding utf8

=cut
